mqtt/mqtt-broker/src/snapshot.rs

```diff
@@ -62,13 +62,35 @@ impl<P> Snapshotter<P> {
 
 impl<P> Snapshotter<P>
 where
-    P: Persist,
+    P: Persist + Send + 'static,
 {
-    pub async fn run(mut self) -> P {
-        while let Some(event) = self.events.recv().await {
+    pub async fn run(self) -> Result<P, Error> {
+        let (tx, rx) = tokio::sync::oneshot::channel::<()>();
```

Comment:
I couldn't read this method. It's rather complicated involving oneshot channel, separate thread, async tasks, and external channel. Can we make it simpler or more clear?

Reply:
I don't think I can simplify it. I can add comments to explain what it is doing?

We want two things:
1. this `run` function needs to be `async` so that it can be joined with the other futures in the `server`
2. We need to handle and propagate any panics on the thread to the event loop thread so that we properly exit, instead of continuing to run with a dead thread.

The `oneshot` is to make this function `async`. The `handle.join()` is to propagate any panics.